src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp

```diff
@@ -261,6 +264,90 @@ ov::npuw::LLMInferRequest::LLMInferRequest(const std::shared_ptr<ov::npuw::LLMCo
+        }
+    }
+
+    // Create all variant requests and share past KV tensors
+    for (size_t i = 0; i < compiled_model->m_generate_compiled_variants.size(); ++i) {
+        std::shared_ptr<ov::IAsyncInferRequest> generate_request;
+
+        if (i == compiled_model->m_generate_compiled_variants.size() - 1) {
+            // Use the already created largest variant
+            generate_request = largest_generate_request;
+        } else {
+            // Create smaller variant
+            generate_request = compiled_model->m_generate_compiled_variants[i]->create_infer_request();
+
+            // Share past KV tensors from the largest variant
+            for (const auto& input_port : generate_request->get_compiled_model()->inputs()) {
+                const auto& input_name = input_port.get_any_name();
+                if (input_name.find(layer_names::past_key_values) != std::string::npos) {
+                    if (largest_past_kv_tensors.find(input_name) != largest_past_kv_tensors.end()) {
+                        auto largest_tensor = largest_past_kv_tensors[input_name];
+                        auto small_shape = input_port.get_shape();
+
+                        // Wrap the largest tensor's data pointer with smaller shape
+                        auto shared_tensor = ov::SoPtr<ov::ITensor>(
+                            ov::make_tensor(input_port.get_element_type(), small_shape, largest_tensor->data()),
+                            nullptr);
+
+                        generate_request->set_tensor(input_port, shared_tensor);
+                    }
+                }
+            }
+        }
+
+        m_generate_requests.push_back(generate_request);
+
+        // Build input/output ports mapping for this variant
+        std::unordered_map<std::string, ov::Output<const ov::Node>> variant_in_ports;
+        std::unordered_map<std::string, ov::Output<const ov::Node>> variant_out_ports;
+
+        for (const auto& input_port : generate_request->get_compiled_model()->inputs()) {
+            variant_in_ports.emplace(input_port.get_any_name(), input_port);
+        }
+        for (const auto& output_port : generate_request->get_compiled_model()->outputs()) {
+            variant_out_ports.emplace(output_port.get_any_name(), output_port);
+        }
+
+        m_generate_variant_in_ports.emplace(generate_request, std::move(variant_in_ports));
+        m_generate_variant_out_ports.emplace(generate_request, std::move(variant_out_ports));
+    }
+
+    // Set default to the largest variant for backward compatibility
+    m_kvcache_request = m_generate_requests.back();
+}
+
+std::shared_ptr<ov::IAsyncInferRequest> ov::npuw::LLMInferRequest::select_generate_request(int64_t num_tokens) {
+    // Find the smallest variant that can accommodate expected token count
+    for (size_t i = 0; i < m_kvcache_sizes.size(); ++i) {
+        if (num_tokens <= m_kvcache_sizes[i]) {
```

Comment:
How does it take into account configurable `min_response_len`? What is prompt will be equal to `1024 + (min_response_len - 2)`, will we choose `1024 + min_response_len` variant then and stop generation after the second token??

Reply:
However, what if user says to use `MAX_PROMPT_LEN: 2048` and `MIN_RESPONSE_LEN: 128`.
But he passed prompt of 1024 + 126, so we fit into the `1024 + 128` model's variant. Then we will stop generation on the second token, but that is not what user meant?

Reply:
Thanks a lot!